### DISCUSSION TITLE
Mocking a plugin

### DISCUSSION THREAD

**Discussion by rhettjay** [fastify/fastify#4927] [Q&A] [answered]
In some instances it's not entirely possible or desired to test a plugin. For instance, say I have an authentication or database plugin as in this [semi-minimal repo](https://github.com/rhettjay/min-repo). However, if this code is going to be deployed through a ci/cd pipleline it's possible that during deployment the code could no longer be within the network where accessing the resources it needs (like an auth service) is possible. During the ci/cd process the automated tests would then fail. Or if I'm using another framework like jest or ava which don't re-use global env variables. It feels like I'm doing something wrong or approaching this in the wrong way. Either I have to write a new server which is built with stubs for each test (or a test server), or I have to stub everything in every test.

Has anyone encountered similar issues to this, if so what approach did you take?

auth-plugin.js
```javascript
export default fp(async fastify, options) => {
  const authenticator = new Authenticator(options)
  async function JwtHandler (request, reply) {
  request.caller = await authenticator.authenticate(request.headers)
  }
  fastify.addHook('onRoute', (route) => {
  route.preValidation == null ? [JwtHandler] : [route.preValidation, JwtHandler]
  })
}
```

app.js
```javascript
export async function app () {
  const fastify = await Fastify()
  
  await fastify.register(auth-plugin, options)
  /*could be any number of plugins like a db, etc...*/
  
  return fastify
}
```

index.js
```javascript
async function run () {
  const fastify = await app()
  await fastify.listen({
    host: '0.0.0.0'
    port: 1234
  })
}

run() 
```

test.js
```javascript
import { app } from '../app'
import { FastifyInstance } from 'fastify'

describe('health route unit test', () => {
  let fastif

  beforeAll(async () => {
    fastify = await app()
  })
  beforeEach(() => {
    jest.resetAllMocks()
  })

  describe('GET hello world', () => {
    test('test health check response', async () => {
      const res = await fastify.inject({
        url: '/'
      })
      expect(res.statusCode).toEqual(200)
    })
  })
})
```

**Comment by Uzlopak** [comment, marked as answer]
My personal opinion:

If I need a database for my plugin i spawn a database, if possible. E.g. using mongodb-memory-server. Once I even compiled a redis server in a docker container so that it works locally and in docker and added it to the repository and then started an instance and run my tests against it. 
This is not feasible for db's like BigQuery, but that is probably also possible to mock with a sqllite 

External services like auth systems usually use http as protocol dont change that often their responses. 
You could store reponses from the original endpoint in a fixtures folder and use the npm package `nock` to replay those in your test.

My favorite approach is: you take your favorite http client (axios, fetch, request, undici etc.) and pass it as an option to your plugin (dependency injection). 
If the option is not set you fallback to the original package. 
Then you can write a wrapper around that http client. That wrapper can now be as powerful as you want it to be. So the wrapper should of course be able to make the request against the real endpoint. If that is done, you can extend it to store requests and responses, or only responses in a fixtures folder. 

Then you can run your tests like using snapshots with tap or jest. 
In production your solution uses the real requests and responses. When you are running the tests locally you can instruct your wrapper to (over)write the requests and responses into your fixtures folder. In the ci/cd plattform you check if the CI Env variable is set, and if so use the payloads in the fixtures folders to replay them against your http requests. 

So you also commit the fixtures and if they change while running the tests locally you can also see the differences and check if the payloads are changed and commit them also.

**Reply by rhettjay** [reply]
Interesting approach, thanks for the ideas @Uzlopak!

**Reply by rhettjay** [reply]
I'm marking this as answered. I took a note out of @Uzlopak's book in addition to implementing a series of mocks and dependency injections utilizing those mocks. Instead of importing the server, I instantiate a new server for each test and register the plugin with the desired mock functionality being injected there.

**Comment by shleder** [comment]
This thread’s discussion of fixture-based replay for plugins is exactly the kind of consumer scenario we want to test independently. ReproProof is a supplemental receipt for a deterministic Node/Fastify reproduction, not a replacement for Fastify’s mocking or fixture guidance.

For a disposable/reviewed checkout, the smallest trial is:

```sh
node --version
npx --yes reproproof@0.4.0 capture \
  --expect-exit 1 \
  --expect-match "failure marker" \
  -- node path/to/repro.js
npx --yes reproproof@0.4.0 verify --no-replay --json reproproof.json
```

Please run `--no-replay` first and only replay a receipt from a trusted checkout. A maintainer should review the Markdown by hand; share only sanitized observations or a small reviewed excerpt—never raw JSON, secrets, private paths, or unredacted logs.

The independent consumer-trial context is tracked here: https://github.com/shleder/reproproof/issues/18

No adoption commitment is requested; a short note on whether the supplemental runtime, lockfile, and exit context helps review one real case would be enough.
